core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -663,47 +666,53 @@ private void mergeFreeEntries(List<FileEntry> entries) {
    @Override
    public void purge(Executor threadPool, final PurgeListener task) {
 
-      threadPool.execute(new Runnable() {
-         @Override
-         public void run() {
-            long now = timeService.wallClockTime();
-            List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<KeyValuePair<Object, FileEntry>>();
-            synchronized (entries) {
-               for (Iterator<Map.Entry<K, FileEntry>> it = entries.entrySet().iterator(); it.hasNext(); ) {
-                  Map.Entry<K, FileEntry> next = it.next();
-                  FileEntry fe = next.getValue();
-                  if (fe.isExpired(now)) {
-                     it.remove();
-                     entriesToPurge.add(new KeyValuePair<Object, FileEntry>(next.getKey(), fe));
-                  }
+      ExecutorAllCompletionService eacs = new ExecutorAllCompletionService(threadPool);
+      Future<Void> future = eacs.submit(() -> {
+         long now = timeService.wallClockTime();
+         List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<KeyValuePair<Object, FileEntry>>();
+         synchronized (entries) {
+            for (Iterator<Map.Entry<K, FileEntry>> it = entries.entrySet().iterator(); it.hasNext(); ) {
+               Map.Entry<K, FileEntry> next = it.next();
+               FileEntry fe = next.getValue();
+               if (fe.isExpired(now)) {
+                  it.remove();
+                  entriesToPurge.add(new KeyValuePair<Object, FileEntry>(next.getKey(), fe));
                }
             }
+         }
 
-            resizeLock.readLock().lock();
-            try {
-               for (Iterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.iterator(); it.hasNext(); ) {
-                  KeyValuePair<Object, FileEntry> next = it.next();
-                  FileEntry fe = next.getValue();
-                  if (fe.isExpired(now)) {
-                     it.remove();
-                     try {
-                        free(fe);
-                     } catch (Exception e) {
-                        throw new PersistenceException(e);
-                     }
-                     if (task != null) task.entryPurged(next.getKey());
+         resizeLock.readLock().lock();
+         try {
+            for (Iterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.iterator(); it.hasNext(); ) {
+               KeyValuePair<Object, FileEntry> next = it.next();
+               FileEntry fe = next.getValue();
+               if (fe.isExpired(now)) {
+                  it.remove();
+                  try {
+                     free(fe);
+                  } catch (Exception e) {
+                     throw new PersistenceException(e);
                   }
+                  if (task != null) task.entryPurged(next.getKey());
```

Comment:
I had thought about that.  My worry is that the listener may not do anything, so it would end up causing a context switch with no benefit.  I am thinking I can submit them in batches using a parallel stream though.  Check it out after I resubmit to let me know what you think.